rust-runtime/aws-smithy-mocks-experimental/src/lib.rs

```diff
@@ -75,6 +75,46 @@ macro_rules! mock {
     };
 }
 
+// This could be obviated by a reasonable trait, since you can express it with SdkConfig if clients implement From<&SdkConfig>.
+
+/// `mock_client!` macro produces a Client configured with a number of Rules and appropriate test default configuration.
+///
+/// # Examples
+/// **Create a client that uses a mock failure and then a success**:
+/// ```rust
+/// use aws_sdk_s3::operation::get_object::{GetObjectOutput, GetObjectError};
+/// use aws_sdk_s3::types::error::NoSuchKey;
+/// use aws_sdk_s3::Client;
+/// use aws_smithy_types::byte_stream::ByteStream;
+/// use aws_smithy_mocks_experimental::{mock_client, mock, RuleMode};
+/// let get_object_happy_path = mock!(Client::get_object)
+///   .match_requests(|req|req.bucket() == Some("test-bucket") && req.key() == Some("test-key"))
+///   .then_output(||GetObjectOutput::builder().body(ByteStream::from_static(b"12345-abcde")).build());
+/// let get_object_error_path = mock!(Client::get_object)
+///   .then_error(||GetObjectError::NoSuchKey(NoSuchKey::builder().build()));
+/// let client = mock_client!(aws_sdk_s3, RuleMode::Sequential, &[&get_object_error_path, &get_object_happy_path]);
+/// ```
+#[macro_export]
+macro_rules! mock_client {
+    ($aws_crate: ident, $rules: expr) => {
+        mock_client!($aws_crate, $crate::RuleMode::Sequential, $rules)
+    };
+    ($aws_crate: ident, $rule_mode: expr, $rules: expr) => {{
+        let mut mock_response_interceptor =
+            $crate::MockResponseInterceptor::new().rule_mode($rule_mode);
+        for rule in $rules {
+            mock_response_interceptor = mock_response_interceptor.with_rule(rule)
+        }
+        $aws_crate::client::Client::from_conf(
+            $aws_crate::config::Config::builder()
+                .with_test_defaults()
+                .region($aws_crate::config::Region::from_static("us-east-1"))
+                .interceptor(mock_response_interceptor)
```

Comment:
I wonder if it's confusing if we accept the package root instead of the `Client` type here. Was this required to get access to the `Config` struct? Maybe at some point we can expose the builder from the client to fix this issue.

Maybe we should add `::builder()` to client that is the config builder or something...